File: lib/arTTY/art.rb
# encoding: utf-8

require "hilighter"

class ArTTY::Art
    attr_reader :pixels
    attr_accessor :legend
    attr_accessor :name
    attr_accessor :sysinfo

    def draw
        offset = nil
        out = ""
        pixels = @pixels
        sysinfo = Array.new
        if (@sysinfo)
            offset = @pixels[0] ? @pixels[0].length + 2 : 0
            sysinfo = @sysinfo.info
        end

        if ((pixels.length % 2) != 0)
            filler = " " * pixels[0].length
            pixels.insert(0, filler)
        end

        while (!pixels.empty?)
            out += " "
            top = pixels.delete_at(0)
            bottom = pixels.delete_at(0)

            top += " " while (top.length < bottom.length)
            bottom += " " while (top.length > bottom.length)

            top.chars.zip(bottom.chars).each_with_index do |map, i|
                t = map[0].strip
                b = map[1].strip

                if (
                    (t.empty? && b.empty?) ||
                    (!t.empty? && !@legend.include?(t)) ||
                    (!b.empty? && !@legend.include?(b))
                )
                    out += " "
                    next
                end

                if (t.empty?)
                    c = "▄".send(@legend[b])
                elsif (b.empty?)
                    c = "▄".send(@legend[t]).swap
                else
                    c = "▄".send(@legend[b]).send("on_#{@legend[t]}")
                end

                out += c
            end

            if (offset)
                info = sysinfo.delete_at(0)
                if (info)
                    filler = offset - top.length
                    out += " " * filler if (filler > 0)
                    out += info
                else
                    offset = nil
                end
            end

            out += "\n"
        end

        if (offset)
            loop do
                info = sysinfo.delete_at(0)
                break if (info.nil?)
                out += " " * (offset + 1)
                out += "#{info}\n"
            end
        end

        return out
    end

    def height
        return @json["height"]
    end

    def initialize(file = nil)
        if (!file.nil?)
            @file = Pathname.new(file).expand_path
            @json = JSON.parse(File.read(@file))
        else
            @file = nil
            @json = Hash.new
        end

        @debug = false
        @legend = @json["legend"] || Hash.new
        @name = @json["name"] || "none"
        @pixels = @json["pixels"] || Array.new
        @sysinfo = nil
    end

    def to_s
        return draw
    end

    def width
        return @json["width"]
    end
end

File: art/art.go
package art

import (
	"encoding/json"
	"os"
	"strings"

	"github.com/mjwhitta/errors"
	hl "github.com/mjwhitta/hilighter"
	"github.com/mjwhitta/pathname"
	"github.com/mjwhitta/sysinfo"
)

// Art is a struct containing the necessary data to convert an ArTTY
// JSON file to pixel art.
type Art struct {
	Height  int               `json:"height"`
	Legend  map[string]string `json:"legend"`
	Name    string            `json:"name"`
	Pixels  []string          `json:"pixels"`
	SysInfo *sysinfo.SysInfo  `json:"-"`
	Width   int               `json:"width"`
}

// New is a constructor for the Art type that takes an optional
// filename.
func New(paths ...string) (*Art, error) {
	var b []byte
	var e error
	var fn string

	if len(paths) == 0 {
		return &Art{}, nil
	}

	fn = pathname.ExpandPath(paths[0])

	if b, e = os.ReadFile(fn); e != nil {
		return nil, errors.Newf("failed to read %s: %w", fn, e)
	}

	return NewFromJSON(b)
}

// NewFromJSON is a constructor for the Art type that takes a JSON
// blob.
func NewFromJSON(b []byte) (*Art, error) {
	var a Art

	if e := json.Unmarshal(b, &a); e != nil {
		return nil, errors.Newf("failed to read JSON: %w", e)
	}

	return &a, nil
}

// String will convert the Art struct to a string for fmt's print
// functions.
func (a *Art) String() string {
	var bottom string
	var filler string
	var info []string
	var line string
	var out []string
	var top string

	if a.Name == "" {
		if a.SysInfo != nil {
			return a.SysInfo.String()
		}

		return ""
	}

	if len(a.Pixels) > 0 {
		filler = strings.Repeat(" ", len(a.Pixels[0]))
	}

	if (len(a.Pixels) % 2) != 0 {
		a.Pixels = append(a.Pixels, filler)
	}

	if a.SysInfo != nil {
		info = strings.Split(a.SysInfo.String(), "\n")
	}

	for i, row := range a.Pixels {
		// Parse 2 rows at a time
		if top == "" {
			top = row
			continue
		}
		bottom = row

		// Make same length
		for len(top) < len(bottom) {
			top += " "
		}
		for len(bottom) < len(top) {
			bottom += " "
		}

		// Create line
		line = " " + a.zipRow(top, bottom)

		// Append SysInfo, if any
		if int(i/2) < len(info) {
			line += " " + info[int(i/2)]
		}

		out = append(out, line)

		// Reset
		top = ""
	}

	for i := int(len(a.Pixels) / 2); i < len(info); i++ {
		line = " " + filler + " " + info[i]
		out = append(out, line)
	}

	return strings.Join(out, "\n")
}

func (a *Art) zipRow(top string, bottom string) (line string) {
	var b string
	var bempty bool
	var bfound bool
	var t string
	var tempty bool
	var tfound bool

	for i := 0; i < len(top); i++ {
		t = string(top[i])
		b = string(bottom[i])

		bempty = (b == " ")
		tempty = (t == " ")
		_, bfound = a.Legend[b]
		_, tfound = a.Legend[t]

		if (tempty && bempty) ||
			(!tempty && !tfound) ||
			(!bempty && !bfound) {
			line += " "
			continue
		}

		if tempty {
			line += hl.Hilight(a.Legend[b], "▄")
		} else if bempty {
			line += hl.Hilights(
				[]string{a.Legend[t], "swap"},
				"▄",
			)
		} else {
			line += hl.Hilights(
				[]string{a.Legend[b], "on_" + a.Legend[t]},
				"▄",
			)
		}
	}

	return
}
